tools/raisinbread_refresh.php

```diff
@@ -0,0 +1,316 @@
+
+require_once 'generic_includes.php';
+$info = <<<INFO
+This script is used by LORIS developers to DELETE DATA and replace
+them with new test data.
+
+
+INFO;
+
+echo $info;
+
+echo "\e[0;31m*** Executing this script will result in the LOSS OF DATA! ***\e[0m\n";
+
+$cwd = getcwd();
+if (substr_compare($cwd, 'tools', mb_strlen($cwd) - mb_strlen('tools')) !== 0) {
+    die('Please run this script from the tools/ directory.' . PHP_EOL);
+}
+
+try {
+    $dbInfo = $config->getSettingFromXML('database');
+
+    if (! $config->getSetting('dev')['sandbox']) {
+        throw new \LorisException(
+            "Config file indicates that this is not a sandbox. Aborting to " .
+            "prevent accidental data loss." . PHP_EOL
+        );
+    }
+    $dbname = $dbInfo['database'];
+    $host   = $dbInfo['host'];
+    // The admin user and password should be configured to a user with DROP
+    // and CREATE permissions for the database.
+    $username = $dbInfo['adminUser'];
+    $password = $dbInfo['adminPassword'];
+
+    $urlConfigSetting  = $config->getSetting('url');
+    $baseConfigSetting = $config->getSetting('base');
+    $hostConfigSetting = $config->getSetting('host');
+
+    echo <<<CONFIRMATION
+    Please type the database name `$dbname` to confirm you wish to drop tables
+    and import test data: 
+CONFIRMATION;
+
+    $input = trim(fgets(STDIN));
+    if ($input !== $dbname) {
+        die('Input did not match database name. Exiting.');
+    }
+
+} catch (\DatabaseException $e) {
+    echo 'Could not connect to the database in the Config file.' .
+        'It\'s possible that the LORIS tables have already been dropped.' .
+        PHP_EOL;
+    echo 'Continue attempting to install Raisinbread database? (y/N)' . PHP_EOL;
+    $input = trim(fgets(STDIN));
+
+    if (mb_strtolower($input) !== 'y') {
+        die;
+    }
+    echo PHP_EOL . 'Please enter the name of your database:' . PHP_EOL;
+    $dbname = trim(fgets(STDIN));
+} catch (\LorisException $e) {
+    die($e->getMessage());
+} catch (Exception $e) {
+    die("Could not load project/config.xml");
+}
+
+
+$mysqlCommand = <<<CMD
+mysql -A $dbname
+CMD;
+
+// Test whether a connection to MySQL is possible via a MySQL config file.
+// If not, read DB information from the config file. This method is not as
+// preferable because it generates MySQL warnings due to the password being
+// supplied via a command-line argument.
+exec($mysqlCommand . ' -e "show tables;" 2>&1 1>/dev/null', $output, $status);
+if ($status != 0) {
+    $mysqlCommand = <<<CMD
+mysql -A "$dbname" -u "$username" -h "$host" -p$password
+CMD;
+}
+
+// Drop tables
+echo printHeader('Dropping LORIS tables.');
+dropTables();
+
+// Print the names of remaining tables, if any. Some tables may remain if they
+// have been created at a different time during development and are not
+// deleted by the above script. Issues with e.g. foreign keys can be a source
+// of error and so it is useful to know what tables remain.
```

Comment:
I'm reviewing the script from the top-down so maybe I'll be elucidated later, but I don't really understand this comment. What's the difference between a "LORIS table" and a "remaining table"? A different time during development of what?

Reply:
@driusan Instruments table. These are not deleted by the 9999-99-99 script and prevent come LORIS table to be deleted due to foreign keys.

Reply:
@PapillonMcGill not exactly. There is a 9999 deletion script for instruments

The "remaining tables" are basically tables that were created at a later stage of development. 
Imagine PR 1 is open and still being reviewed and PR 5 adds a new table `BLA`, I come and test PR 5 so I create the `BLA` table on my database. then I try to test PR 1, but this tool will not delete `BLA` because the 9999 script on PR 1 does not contain a drop statement for `BLA`. Now imagine `BLA` has a foreign key reference on candidate, then not only will `BLA` still be there but it will cause this tool to fail entirely when deleting the candidate table.

So the tool will need to be flexible to be able to work on branches at different points in time

Reply:
@ridz1208 Then why when I tried it, Instruments table where not deleted and prevented some LORIS tables to be deleted due to foreign keys?

Reply:
@PapillonMcGill line 194 is the deletion for instruments

and if you have your own it's not technically the RB dataset... so yes this would help you, but that's not why it is necessary and finally there are usually not foreign key constraints between instruments and the rest of core loris tables.

Reply:
@ridz1208 It didn't ran or it tried but failed for any reason without any error message when I tried to test this PR.

Reply:
Yeah Rida's explanation is correct. This part of the code is for two cases:

1. You've installed a table from another PR that isn't installed on the current branch (and as a corollary is not included in the deletion scripts). E.g. if I have `biobank` tables in my DB, this makes sure they get deleted.
2. You have non-standard tables that you've created. These might have FKs to core tables. This will cause the script to fail when it tries to delete core tables.

This part of the code should print and optionally delete all tables in your DB that aren't part of Raisinbread on the current branch.

Reply:
@ridz1208 A good example of why you have extra table and foreign key, is when testing PRs. Test a new PR that add new tables (and data using foreign keys to existing data) then switch to another PR and use this script to reset your DB. You have some tables that are still unknown to the 9999-99-99 script in the later PR.

Reply:
lol @PapillonMcGill thats litterally my explanation from 7 days ago

https://github.com/aces/Loris/pull/5260#discussion_r368546801

Reply:
@ridz1208 from what I remembered, you seam to find that as not needed.

Reply:
I dont beleive I ever found it not needed no. thus the `need` key work in my comment
```
So the tool will need to be flexible to be able to work on branches at different points in time
```

Reply:
Either way that feature exists so we should be good!